### ISSUE TITLE
`string` is seen as `record` but when trying to use the `record` the `string` is recognised

### ISSUE THREAD

**Issue by theAkito** [nushell/nushell#9107]
### Describe the bug

```nu
let dir_output = "output_posts"
let file_output_prefix = "posts_"
let id_admin = 1
let api_key = 'api_key'
let root_url = 'https://mydomain.eu/api/'
let api_posts = $"($root_url)posts"
let header_auth_val = [ Authorization $"Bearer ($api_key)" ]

def retrieveDiscussionRoot [url: string = $api_posts] {
  print "retrieveDiscussionRoot"
  (
    print $url;
    let p = (http get -H $header_auth_val $url);
    print "retrieveDiscussionRoot $p";
    print $p;
    print ($p | describe);
    echo $p
    | from json
  )
}

def retrieveDiscussionAttrs [url: string = $api_posts] {
  (
    http get -H $header_auth_val $url
    | from json
    | get data
    # | par-each { |row| $row.attributes }
  )
}

def getDiscussionAttrs [] {
  (
    $in
    | get data
    # | par-each { |row| $row.attributes }
  )
}

let posts = (
  mut root = (retrieveDiscussionRoot);
  mut result = ($root | getDiscussionAttrs);
  mut link_next = ($root | get links?.next?);
  loop {
    $link_next = ($root | get links?.next?)
    if $link_next == null { break }
    print 1
    if ($root | getDiscussionAttrs | length) == 20 {
      print 2
      $result = ($result | append (retrieveDiscussionAttrs $link_next))
      sleep 20sec
    } else {
      break
    }
    print 3
    print $link_next
    print 3a
    print $root
    print 3b
    print $result
    # print 3c
    # print $link_next
    $root = (retrieveDiscussionRoot $link_next)
  };
  $result
)

mkdir $dir_output | ignore

{
  date_rfc2822: (date now | date format)
  posts_count: ($posts | length)
  posts: ($posts | group-by relationships)
} | to json | save $"($dir_output)/($file_output_prefix)(date now | date format "%Y-%m-%dT%H:%M:%S").json"
```

Running this script results in the following error.

```
Error: nu::shell::cant_convert

  × Can't convert to string.
    ╭─[/path/to/posts.nu:34:1]
 34 │     print ($p | describe);
 35 │     echo $p
    ·          ─┬
    ·           ╰── can't convert record<user: record<bla bla bla>> to string
 36 │     | from json
    ╰────
```

The problem is basically, that the GET request retrieves a string, which is interpreted by Nu as an already parsed record, but when you try to use it as a table, then it complains, that it is a string, which it actually is.
Now, when you treat the result as a string, which would be correct, then it complains that it is already a parsed record.

### How to reproduce

1. Take the script above.
2. Let the GET request to the URL you provide return a long JSON with lots of variable data types & corresponding values.
3. Run the script.
4. See, how it interprets the `string` as a `record`, but when you try to use the supposed `record`, it then understands it, correctly, as a `string`.

### Expected behavior

assert `string` == `string`
assert `record` == `record`

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.79.0                                   |
| branch             |                                          |
| commit_hash        | a1b72611215dbfca257351003204a80c83859e05 |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.66.1 (90743e729 2023-01-10)      |
| rust_channel       | 1.66.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.66.1 (ad779e08b 2023-01-10)      |
| build_time         | 2023-04-25 20:26:29 +00:00               |
| build_rust_channel | release                                  |
| features           | default, zip                             |
| installed_plugins  |                                          |

### Additional context

* Similar issue about type mistakes: https://github.com/nushell/nushell/issues/7699

I assume, this is related to the feature request, requesting that `| from json` shouldn't be necessary anymore, if the issue was already worked on. I cannot find the particular issue, at this moment.

**Comment by amtoine** [comment]
@theAkito 
thanks for reporting this :yum: 

can you try to provide a smaller script and a sample of data so that it will be easier to investigate?
that would be amazing :sparkles:

**Comment by theAkito** [comment]
> can you try to provide a smaller script and a sample of data so that it will be easier to investigate?

When trying to load the entire original JSON API output from a local file, rather than from the remote API, it complains about an `"invalid escape"`.

Perhaps the third party backend I am using does not encode non-ASCII symbols well. The error pointed to a place with escaped Unicode symbols. 🤔

Tried to load it in another language & indeed, getting Syntax error at the same place. Apparently, the third party backend does not encode those symbols well.

After removing this particular symbol (tons of other Unicode symbols were apparently properly escaped), the parsing works. In a local example script for reproduction purposes, the parsing went fine.

Still, Nu is buggy, anyway.
1. Why is there such a weird error, rather than a syntax error? This is one Nu bug.
2. Why is the error different in a local example, where I am loading the payload from a local file rather than from a remote API call, by using `http get`? This is inconsistent & again points to `http get` making it worse. This is another Nu bug.

**Comment by fdncred** [comment]
It may also be that double quotes interpret escapes. Single quotes don't.

**Comment by fdncred** [comment]
@theAkito we need something reproducible in order to debug and try to fix this. as detailed as your post is, i'm not sure how to test it. Maybe a sample json file is all you need?

**Comment by theAkito** [comment]
@fdncred

It's really hard to get a minimal sample without cutting off too much.

I think however, I have something that reproduces the issue.

Try & tell me, whether it works, please.

```json
{"canEdit":true,"canEditCse,"must_accept":false}
```

This is a tiny snippet of the JSON produced by the third party service. Needless to say, the JSON issue itself is a bug to be reported on their end.

**Comment by theAkito** [comment]
> @theAkito I'm not sure what to do with this json snippet to reproduce the error.

You need to run the script, as seen in the original issue. I still have not found a way to reproduce it with a small, quick command, like `open`, because then you'll just get a plain old parsing error, which is correct.

You need to have a backend, sending the json & then you need to process it like this.

```nu
def retrieveDiscussionRoot [url: string = $api_posts] {
  print "retrieveDiscussionRoot"
  (
    print $url;
    let p = (http get -H $header_auth_val $url);
    print "retrieveDiscussionRoot $p";
    print $p;
    print ($p | describe);
    echo $p
    | from json
  )
}
```

Which is obviously my attempt at trying to debug this problem, so you just need this.

```nu
http get http://127.0.0.1/api/faulty | from json
```

Now, set up a backend at `http://127.0.0.1/api/faulty`. Done.

**Comment by fdncred** [comment]
i won't be able to do that though. I can do this
```
let p = (open -r test.json)
```
and now the invalid json is in `$p` but it's not parsed yet. and you can do `print $p` but when you do `from json` it will puke but with a different error message and don't see it saying string as record.
```
echo $p | from json
Error:
  × Error while parsing JSON text
   ╭─[entry #56:1:1]
 1 │ echo $p | from json
   ·      ─┬
   ·       ╰── error parsing JSON text
   ╰────

Error:
  × Error while parsing JSON text
   ╭────
 1 │ {"canEdit":true,"canEditCse,"must_accept":false}
   ·                               ▲
   ·                               ╰── "expected `:`" at line 1 column 30
   ╰────
```

**Comment by melMass** [comment]
> {"canEdit":true,"canEditCse,"must_accept":false}

This is not valid json

**Comment by theAkito** [comment]
> > {"canEdit":true,"canEditCse,"must_accept":false}
> 
> This is not valid json

Yeah, no kidding... 🤦🏻‍♀️ 🙄

**Comment by melMass** [comment]
> Yeah, no kidding... 🤦🏻‍♀️ 🙄

Hey chill, this is you:

> This is a tiny snippet of the JSON produced

**Comment by theAkito** [comment]
> > Yeah, no kidding... 🤦🏻‍♀️ 🙄
> 
> Hey chill, this is you:
> 
> > This is a tiny snippet of the JSON produced



> This is a tiny snippet of the JSON produced by the third party service. **Needless to say, the JSON issue itself is a bug to be reported on their end.**

Important part is quoted in bold.

____

This issue is about how Nu handles an _invalid_ JSON (resulting from an HTTP request). It should complain, that it is invalid, instead of seemingly converting it into a Limbo between `record` & `string`.

**Comment by melMass** [comment]
Tbh it's quite hard to even try. But I have seen a couple of times API returning either what you describe or even multiple objects but without delimiters besides their brackets.
What I'm not sure to get is if the issue lies in `from json` or `http get`.

What is the output of `| describe` after each?

**Comment by cultpony** [comment]
I've ran into this issue today, it is quite annoying to debug since nushell seems to have trouble pinning down where the error comes from.

The issue on my end is caused by relatively large JSON objects (Kubernetes objects containing other Kubernetes objects with history), so that might be a part of the problem.

A workaround is to use JQ to extract parts of the JSON and parse the smaller segments individually, then reassemble the entire object afterwards (and I also used this step to eliminate some unneeded metadata fields that were quite large)

**Comment by fdncred** [comment]
@cultpony do you have a small reproducible sample that we can test and debug with?

**Comment by cultpony** [comment]
Not really, everytime I reduce the code to become reproducible it goes away, but I haven't tried that much. In my case, about 6 or 7 stackframes of nested (par-)each calls or so. I've got an upcoming free day, I'll see if I can make something reproducible then.

**Comment by Bahex** [comment]
There has been a lot of changes since this issue was opened, including large refactors of both error handling and the type system so I'm closing this issue.

If anyone can reproduce the problem please share the example here and I'll reopen the issue.
